plugins/bufferlines/barbar.nix

```diff
@@ -69,249 +24,592 @@ let
 
-    minimumPadding =
-      helpers.defaultNullOpts.mkInt 1
-        "Sets the minimum padding width with which to surround each tab";
+      diagnostics = mkOption {
+        type = types.submodule {
+          freeformType = with types; attrsOf anything;
+          options =
+            let
+              mkDiagnosticIconOptions = iconDefault: {
+                enabled = helpers.defaultNullOpts.mkBool false ''
+                  Enable showing diagnostics of this `|diagnostic-severity|` in the 'tabline'.
+                '';
+
+                icon = helpers.defaultNullOpts.mkStr iconDefault ''
+                  The icon which accompanies the number of diagnostics.
+                '';
+              };
+            in
+            {
+              "vim.diagnostic.severity.ERROR" = mkDiagnosticIconOptions " ";
+              "vim.diagnostic.severity.HINT" = mkDiagnosticIconOptions "󰌶 ";
+              "vim.diagnostic.severity.INFO" = mkDiagnosticIconOptions " ";
+              "vim.diagnostic.severity.WARN" = mkDiagnosticIconOptions " ";
+            };
+        };
+        apply = helpers.toRawKeys;
+        default = { };
+        description =
+          helpers.defaultNullOpts.mkDesc
+            {
+              "vim.diagnostic.severity.ERROR" = {
+                enabled = false;
+                icon = " ";
+              };
+              "vim.diagnostic.severity.HINT" = {
+                enabled = false;
+                icon = "󰌶 ";
+              };
+              "vim.diagnostic.severity.INFO" = {
+                enabled = false;
+                icon = " ";
+              };
+              "vim.diagnostic.severity.WARN" = {
+                enabled = false;
+                icon = " ";
+              };
+            }
+            ''
+              Set the icon for each diagnostic level.
+              The keys will be automatically translated to raw lua:
+              ```nix
+                {
+                  "vim.diagnostic.severity.INFO".enabled = true;
+                  "vim.diagnostic.severity.WARN".enabled = true;
+                }
+              ```
+              will result in the following lua:
+              ```lua
+                {
+                  -- Note the table keys are not string literals:
+                  [vim.diagnostic.severity.INFO] = { ['enabled'] = true },
+                  [vim.diagnostic.severity.WARN] = { ['enabled'] = true },
+                }
+              ```
+            '';
+      };
+
+      gitsigns =
+        helpers.defaultNullOpts.mkAttrsOf
+          (
+            with types;
+            submodule {
+              freeformType = with types; attrsOf anything;
+              options = {
+                enabled = mkOption {
+                  type = types.bool;
+                  description = helpers.defaultNullOpts.mkDesc true ''
+                    Enables showing git changes of this type.
+                    Requires `|gitsigns.nvim|`.
+                  '';
```

Comment:
`enabled` and `icon` aren't nullable or maybeRaw. Is that intentional?

Why not just use `defaultNullOpts.mkBool true` here, instead of manually using `mkOption` and `mkDesc`?

Reply:
Yes, this is intentional. Each element of this attrs should be a module with those two keys.

Reply:
In `diagnostics` you use `defaultNullOpts` submodule options:

```nix
mkDiagnosticIconOptions = iconDefault: {
  enabled = helpers.defaultNullOpts.mkBool false ''
    Enable showing diagnostics of this `|diagnostic-severity|` in the 'tabline'.
  '';

  icon = helpers.defaultNullOpts.mkStr iconDefault ''
    The icon which accompanies the number of diagnostics.
  '';
};
```

Why is `gitsigns` different?

If it's just that when all sub-options must be non-null, then we could still support raw types via `maybeRaw`.

Unless there's some other reason not to?

Reply:
Either way, I think there should be a comment to explain why the submodule's options aren't `defaultNullOpts` and/or aren't `maybeRaw`

Reply:
Also, if the option _must_ be present then surely `mkDesc true` makes no sense, since logically a required option can't have a plugin-default.

Reply:
Is it now OK @MattSturgeon ?